src/main/java/org/commcare/suite/model/Entry.java

```diff
@@ -157,14 +158,22 @@ public String getDisplayText() {
     }
 
     @Override
-    public Single<String> getTextForBadge(final EvaluationContext ec) {
+    public Single<String> getTextForBadge(EvaluationContext ec) {
+        final EvaluationContext[] abandonableContext =
+                new EvaluationContext[] {ec.spawnWithCleanLifecycle()};
         if (display.getBadgeText() == null) {
             return Single.just("");
         }
         return Single.fromCallable(new Callable<String>() {
             @Override
             public String call() throws Exception {
-                return display.getBadgeText().evaluate(ec);
+                return display.getBadgeText().evaluate(abandonableContext[0]);
+            }
+        }).doOnDispose(new Action() {
```

Comment:
Same needs to go into Menu.java implementation. Better to make a default implementation for both Menu and Entry class at this point.

Reply:
Good point. I'll unify a thread thing for "Evaluating an EC with these semantics."